### DISCUSSION TITLE
Multiple components of same type with different query string for same property

### DISCUSSION THREAD

**Discussion by wimski** [livewire/livewire#7862] [1. Help] [answered]
# Setup
### Component Class
```php
use Livewire\Attributes\Locked;
use Livewire\Component;

class Test extends Component
{
    public string $value = '';

    #[Locked]
    public string $queryParamName;

    /**
     * @return array<string, array<string, string>>
     */
    protected function queryString(): array
    {
        // @see: https://livewire.laravel.com/docs/url#using-the-querystring-method
        // "The query string can also be defined as a method on the component.
        // This can be useful if some properties have dynamic options."
        return [
            'value' => ['as' => $this->queryParamName],
        ];
    }
}
```
### Component Template
```blade
<div>
    <ul>
        <li>value: {{ $value }}</li>
        <li>queryParamName: {{ $queryParamName }}</li>
    </ul>
    <button type="button" wire:click="$set('value', '{{ \Illuminate\Support\Str::random() }}')">set</button>
</div>
```
### Page
```blade
<livewire:test.test queryParamName="foo" />
<livewire:test.test queryParamName="bar" />
```
# Issue
When clicking the button in the first component I would expect the URL query parameters to be `?foo={SOME_STRING}`, but it becomes `?foo={SOME_STRING}&bar={SOME_STRING}`, where both instances of `SOME_STRING` are the `$value` of the first component. Why do both query strings get set and can this be remedied in some way?

**Comment by danie-ramdhani** [comment]
maybe wire:key works?

```blade
<livewire:test.test queryParamName="foo" wire:key="foo" />
<livewire:test.test queryParamName="bar" wire:key="bar" />
```

**Reply by wimski** [reply]
Unfortunately that doesn't make a difference.

**Reply by danie-ramdhani** [reply]
Oh yeah you right... probably bug...

**Comment by joshhanley** [comment, marked as answer]
@wimski yeah that looks like it's a bug. Any chance you could submit a failing test PR? Thanks!

**Reply by wimski** [reply]
I'll definitely see what I can do to help.

**Reply by joshhanley** [reply]
@wimski thanks! Much appreciated. If you get stuck, let me know 🙂

**Reply by wimski** [reply]
@joshhanley should this failing test go in `Livewire\Tests\QueryParamsUnitTest` or `LegacyTests\Browser\QueryString\Test`?

**Reply by joshhanley** [reply]
@wimski actually it should go in `Livewire\Features\SupportQueryString` folder, into either the `UnitTest` or `BrowserTest` file depending on what type of test you will write. I suspect it will need to be a browser test.

**Reply by wimski** [reply]
@joshhanley I followed the [contributing guide](https://livewire.laravel.com/docs/contribution-guide), but running the tests results in this :grimacing: 
```sh
Tests: 891, Assertions: 2026, Errors: 236, Failures: 8, Skipped: 43, Incomplete: 3.
```
The errors seem to fall in three categories:
* `Failed to connect to localhost port 9515`
* `500` errors (both in status assertions and views), could be related to the previous.
* `SQLSTATE[HY000]: General error: 1 no such table: sessions (Connection: testbench, SQL: select * from "sessions" where "id" = ? limit 1)`

Do I need any additional setup to make this work? A dedicated Docker setup in this code-base would've been great.

**Reply by joshhanley** [reply]
@wimski I've no idea how to use docker 😅

I think sometime you need to close all open chrome windows and instances (the very first time you run dusk). After that it should run ok.

**Reply by wimski** [reply]
Closing my Chrome browser didn't make a difference. I'm afraid getting the setup to work is going to take too long for me (atm) :cry:

**Reply by joshhanley** [reply]
@wimski bugger! Just try to write the dusk test best you can, and then submit the PR and mention you couldn't get Dusk running

**Reply by VitorTS** [reply]
I tried getting the project running yesterday and got the same error. The contribution guide is outdated.

**Reply by joshhanley** [reply]
@VitorTS did you get the same error `failed to connect to localhost`?

**Reply by VitorTS** [reply]
I see this:

<img width="1195" alt="image" src="https://github.com/livewire/livewire/assets/26394646/eb138682-7e2f-43fc-bf53-511d685b0a2c">

**Reply by joshhanley** [reply]
@VitorTS thanks. I'll pull Livewire down fresh and see what happens.

**Reply by joshhanley** [reply]
@VitorTS ok yeah I'm getting the same issue as you on a new clone. I will have a look into what is going on

**Reply by VitorTS** [reply]
thank you!

**Reply by joshhanley** [reply]
Ok so what is happening, is Laravel 11 and the corresponding Orchestra Testbench/ Testbench-dusk versions are being installed.

@wimski and @VitorTS can you both try this (with everything already set up the way you have it):

```
composer update laravel/framework:^10 orchestra/testbench:^8 orchestra/testbench-dusk:^8 -W

./vendor/bin/dusk-updater update
```

And then run PHP unit again:

```
./vendor/bin/phpunit
```

Can you let me know how you go? Thanks!

**Reply by joshhanley** [reply]
Yeah there is an issue with Laravel 11. Have submitted a PR #7874 to fix some of the issues. But it'd be safer to install Laravel 10 as per my command above for the time being.

**Reply by wimski** [reply]
This makes no difference in my case. I probably need some more setup than just a PHP CLI container :thinking:

**Reply by joshhanley** [reply]
@wimski bugger 😕 yeah not sure what's happening with your set up

**Reply by wimski** [reply]
@joshhanley It's definitely a Docker vs Dusk/ChromeDriver issue at my end, because only the `Failed to connect to localhost port 9515` error remains, which is due to the browser not being available.

I had a look to see if I could write a test without running it, but I have a hard time coming up with a setup the 'could' work. The bug arises when you run multiple instances of the same component and I haven't seen a test that does something similar. Is that something a test _can_ handle?

After some more investigation it seems to be JS bug, because the incoming query parameters are handled correctly and the `/update` calls all seem correct. JS is just hard to debug (for me at least) in my project, because everything is compiled.

**Reply by joshhanley** [reply]
@wimski yeah I've seen a lot of people have issues running dusk on docker. I wouldn't know the first thing to get docker running let alone get dusk loading on it 😅

Hmm maybe have a look at the nested component tests? https://github.com/livewire/livewire/tree/main/src/Features/SupportNestingComponents

Yeah if you can mostly get a test together that you think should run, just submit it and mention you couldn't get dusk running.

**Reply by wimski** [reply]
https://github.com/livewire/livewire/pull/7910 :crossed_fingers:

**Reply by joshhanley** [reply]
@wimski thanks! 🙏
